sdk/helper/random/parser_test.go

```diff
@@ -0,0 +1,711 @@
+package random
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestParse(t *testing.T) {
+	type testCase struct {
+		rawConfig string
+		expected  StringGenerator
+		expectErr bool
+	}
+
+	tests := map[string]testCase{
+		"unrecognized rule": {
+			rawConfig: `
+				length = 20
+				charset = "abcde"
+				rule "testrule" {
+					string = "omgwtfbbq"
+					int = 123
+				}`,
+			expected: StringGenerator{
+				Length:  20,
+				Charset: []rune("abcde"),
+				Rules:   nil,
+			},
+			expectErr: true,
+		},
+
+		"charset restrictions": {
+			rawConfig: `
+				length = 20
+				charset = "abcde"
+				rule "CharsetRestriction" {
+					charset = "abcde"
+					min-chars = 2
+				}`,
+			expected: StringGenerator{
+				Length:  20,
+				Charset: []rune("abcde"),
+				Rules: []Rule{
+					&CharsetRestriction{
+						Charset:  []rune("abcde"),
+						MinChars: 2,
+					},
+				},
+			},
+			expectErr: false,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			actual, err := Parse(test.rawConfig)
+			if test.expectErr && err == nil {
+				t.Fatalf("err expected, got nil")
+			}
+			if !test.expectErr && err != nil {
+				t.Fatalf("no error expected, got: %s", err)
+			}
+
+			if !reflect.DeepEqual(actual, test.expected) {
+				t.Fatalf("Actual: %#v\nExpected:%#v", actual, test.expected)
+			}
+		})
+	}
+}
+
+func TestParser_Parse(t *testing.T) {
+	type testCase struct {
+		registry map[string]ruleConstructor
+
+		rawConfig string
+		expected  StringGenerator
+		expectErr bool
+	}
+
+	tests := map[string]testCase{
+		"empty config": {
+			registry:  defaultRuleNameMapping,
+			rawConfig: "",
+			expected:  StringGenerator{},
+			expectErr: true,
+		},
+		"bogus config": {
+			registry:  defaultRuleNameMapping,
+			rawConfig: "asdf",
+			expected:  StringGenerator{},
+			expectErr: true,
+		},
+		"config with length and charset": {
+			registry: defaultRuleNameMapping,
+			rawConfig: `
+				length = 20
+				charset = "abcde"`,
+			expected: StringGenerator{
+				Length:  20,
+				Charset: []rune("abcde"),
+			},
+			expectErr: false,
+		},
+		"config with negative length": {
```

Comment:
How about a length of `0` test?

Reply:
There is a test elsewhere for this, but I can add one here too.